pkg/apis/coordination/types.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package coordination
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Lease defines a lease concept.
+type Lease struct {
```

Comment:
Have we considered something like `/renew` and `/reassign` subresources, I wonder if these would be more convenient to implement against, vs having to manually set the timestamps and transition count in a client?
Also thoughts on clock skew between clients and the API server?

Reply:
I think this may be done on top of that PR - it's an addition, not a change to this PR.

Reply:
@mtaufen This type of issue was already discussed during the KEP itself:
https://github.com/kubernetes/community/pull/2090

Though perhaps we should expand the alternatives section.

Re. timestamps: Clients really shouldn't be using the persisted timestamps. Components instead should treat observed time changes as triggers and use their own timers locally. In the case of crash, restart, etc., components should be conservative and restart their timers.

Re. custom verbs: I understand the sentiment, but there are a number of challenges with it.
1. I want to avoid business logic in kube-apiserver. It could be performed in admission control, I suppose, but I'd prefer admission control be used for validation and policy enforcement only.
2. I want to avoid custom verbs. We're trying to move towards more consistent CRUD.
3. I want to avoid synchronous verbs. The typical verb would just be `/acquire`, blocking while already leased by another. It could be implemented as an asynchronous verb, in which case it would fail if leased by another. However, in the latter case, the client would need to determine reasonable retry intervals, which suggests it would need to reason about the time-related fields. Additionally, in this case, we want kubelet to acquire the lease as soon as it is able, and we want node controller to observe whether the lease is held, but not to acquire the lease itself. So clients need to reason about most of the fields (and need to properly use resourceVersion).
4. We have a purely client-based implementation, and such an implementation is simpler in aggregate for the system for our expected use cases, which are by Kubernetes components, including by the controller manager (for HA leader election). Due to the support for optimistic concurrency, effectively any resource can act as a lease. For a long time, this functionality has been implemented using annotations.

Similarly, the question of whether to use the etcd lease API was raised, which would analogous to the suggested `/reassign` (grant) and `/renew` (keep alive) verbs:
https://coreos.com/etcd/docs/latest/learning/api.html#lease-api

In addition to creating a new, undesirable dependency on the etcd API, that API doesn't expose enough information for intended use cases, such as the holder of the lease. (I'm also not sure what happens in the case of a crash and restart.)

Note that AcquireTime and LeaseTransitions are really just for debugging.